rust/private/unpretty.bzl

```diff
@@ -0,0 +1,380 @@
+        cc_toolchain,
+        feature_configuration,
+        crate_info,
+        dep_info,
+        build_info,
+    )
+
+    output_groups = {}
+    outputs = []
+
+    for mode in ctx.attr._unpretty_modes[RustUnprettyInfo].modes.to_list():
+        pretty_mode = mode.replace("-", "_")
+        mnemonic = "RustUnpretty{}".format("".join([
+            o.title()
+            for m in pretty_mode.split(",")
+            for o in m.split("_")
+        ]))
+
+        unpretty_out = ctx.actions.declare_file(
+            "{}.unpretty.{}.rs".format(ctx.label.name, pretty_mode.replace(",", ".")),
+            sibling = crate_info.output,
+        )
+
+        output_groups.update({"rust_unpretty_{}".format(pretty_mode.replace(",", "_")): depset([unpretty_out])})
+        outputs.append(unpretty_out)
+
+        rust_flags = []
+        if crate_info.is_test:
+            if getattr(ctx.rule.attr, "use_libtest_harness", True):
+                rust_flags = ["--test"]
+            else:
+                rust_flags = ["--cfg", "test"]
+
+        args, env = construct_arguments(
+            ctx = ctx,
+            attr = ctx.rule.attr,
+            file = ctx.file,
+            toolchain = toolchain,
+            tool_path = toolchain.rustc.path,
+            cc_toolchain = cc_toolchain,
+            feature_configuration = feature_configuration,
+            crate_info = crate_info,
+            dep_info = dep_info,
+            linkstamp_outs = linkstamp_outs,
+            ambiguous_libs = ambiguous_libs,
+            output_hash = determine_output_hash(crate_info.root, ctx.label),
+            rust_flags = rust_flags,
+            out_dir = out_dir,
+            build_env_files = build_env_files,
+            build_flags_files = build_flags_files,
+            emit = ["dep-info", "metadata"],
+            skip_expanding_rustc_env = True,
+        )
+
+        args.process_wrapper_flags.add("--stdout-file", unpretty_out)
+
+        # Expand all macros and dump the source to stdout.
+        # Tracking issue: https://github.com/rust-lang/rust/issues/43364
+        args.rustc_flags.add("-Zunpretty={}".format(mode))
+
+        ctx.actions.run(
+            executable = ctx.executable._process_wrapper,
+            inputs = compile_inputs,
+            outputs = [unpretty_out],
+            env = env,
+            arguments = args.all,
+            mnemonic = mnemonic,
+            toolchain = "@rules_rust//rust:toolchain_type",
+        )
+
+    output_groups.update({"rust_unpretty": depset(outputs)})
+
+    return [
+        OutputGroupInfo(**output_groups),
+    ]
+
+# Example: Expand all rust targets in the codebase.
+#   bazel build --aspects=@rules_rust//rust:defs.bzl%rust_unpretty_aspect \
+#               --output_groups=expanded \
+#               //...
+rust_unpretty_aspect = aspect(
+    implementation = _rust_unpretty_aspect_impl,
+    fragments = ["cpp"],
+    host_fragments = ["cpp"],
+    attrs = {
+        "_cc_toolchain": attr.label(
```

Comment:
It feels like we should maybe extract a "common_attrs_needed_for_compiling" constant somewhere which we can reference from a bunch of places, so that when we add a required attr for collect_deps/collect_inputs/construct_arguments we only need to modify one place?

Reply:
Updated!